### PR TITLE
p2p/net/swarm: dial once at a time

### PR DESCRIPTION
**Pull request by jbenet**
This PR makes swarm dials wait for each other.

### CODE REVIEW

**Comment by jbenet** [comment]
@briantigerchow RFCR?

File: p2p/net/swarm/swarm_dial.go [hunk 1]
```diff
@@ -20,17 +92,53 @@ import (
...
-	for _, c := range cs {
-		if c != nil { // dump out the first one we find
-			return c, nil
+	// this loop is here because dials take time, and we should not be dialing
+	// the same peer concurrently (silly waste). Additonally, it's structured
+	// to check s.ConnectionsToPeer(p) _first_, and _between_ attempts because we
+	// may have received an incoming connection! if so, we no longer must dial.
+	//
+	// dial attempts. we may be doing the dialing. if not, we wait.
+	attempts := 3
```
Comment (briantigerchow) [inline p2p/net/swarm/swarm_dial.go]: const?

File: p2p/net/swarm/swarm_dial.go [hunk 2]
```diff
@@ -20,17 +92,53 @@ import (
...
+			case <-wait: // wait for that dial to finish.
+				continue // and see if it worked (loop), OR we got an incoming dial.
+			case <-ctx.Done(): // or we may have to bail...
+				return nil, ctx.Err()
+			}
+		}
+
+		// ok, we have been charged to dial! let's do it.
+		conn, err = s.dial(ctx, p)
+		s.dsync.Unlock(p)
```
Comment (briantigerchow) [inline p2p/net/swarm/swarm_dial.go]: It wasn't immediately clear that the `conn` is added to the swarm's list of conns in the `dial` method, so I thought this lock might've been released too early.

However, I read through and saw that swarm add conn occurs in dialConnSetup.

(nested state changes are somewhat hard to follow)

but the code looks correct. LGTM

File: p2p/net/swarm/swarm_dial.go (hunk 1 above)
Comment (jbenet) [inline p2p/net/swarm/swarm_dial.go]: sgtm - 25c6ab8

**Comment by briantigerchow** [comment]
@jbenet LGTM

File: p2p/net/swarm/swarm_dial.go [hunk 3]
```diff
@@ -20,17 +95,52 @@ import (
...
+			case <-wait: // wait for that dial to finish.
+				continue // and see if it worked (loop), OR we got an incoming dial.
+			case <-ctx.Done(): // or we may have to bail...
+				return nil, ctx.Err()
+			}
 		}
+
+		// ok, we have been charged to dial! let's do it.
+		conn, err = s.dial(ctx, p)
+		s.dsync.Unlock(p)
```
Comment (briantigerchow) [inline p2p/net/swarm/swarm_dial.go]: It wasn't immediately clear that the conn is added to the swarm's list of conns in the dial method, so I thought this lock might've been released too early.

However, I read through and saw that swarm add conn occurs in dialConnSetup.

(nested state changes are somewhat hard to follow. It's nice to have lots of pure code (and to isolate code containing the side-effects, and to keep the side-effects as close to the surface as possible))

but the code looks correct. LGTM

Comment (jbenet) [inline p2p/net/swarm/swarm_dial.go]: Agreed, i was annoyed by this, and other warts. it isn't easy to lift up, without carrying a lot of logic into this part, which I wanted to keep very light-- only focused on the sync attempts (turns out dialing is hard!). I added a note 8e888b1 which may clear it up for the reader.

**Comment by jbenet** [comment]
ok, thanks! I'll merge when the build greens
